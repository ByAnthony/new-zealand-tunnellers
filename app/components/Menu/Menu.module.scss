@use "app/variables.module.scss";

.menu {
  margin-top: variables.$s-margin;
  transition: opacity 0.5s ease;
  display: flex;
  justify-content: center;
  align-items: center;

  @include variables.component-margin;

  @include variables.tablet {
    margin-top: variables.$m-margin;
  }

  @include variables.desktop {
    margin: 0 auto;
    margin-top: variables.$base-margin;
    max-width: variables.$inner-width;
  }

  .logo {
    display: flex;
    align-items: center;
    height: variables.$base-margin;
    margin-right: variables.$xs-margin;
    transition:
      border-bottom 0.9s ease-in-out,
      border-bottom 0.9s ease-in-out,
      border 0.9s ease-in-out,
      border 0.9s ease-in-out,
      background-color 0.9s ease-in-out,
      background-color 0.9s ease-in-out;

    img {
      height: 30px;
    }
  }

  .search-form-container {
    position: relative;
    width: 100%;
  }

  .search-form {
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: variables.$base-margin;
    width: 100%;
    border: 1px solid variables.$primary-light-color;
    border-radius: variables.$border-radius;
    background-color: variables.$primary-medium-color;

    input {
      width: 100%;
      box-sizing: border-box;
      width: 100%;
      height: variables.$base-margin;
      padding: 0 variables.$xs-margin;
      font-family: variables.$main-font;
      font-weight: 300;
      font-size: 1rem;
      color: variables.$text-color;
      border: none;
      border-top-left-radius: variables.$border-radius;
      border-bottom-left-radius: variables.$border-radius;
      background-color: variables.$primary-medium-color;
      transition:
        border-bottom 0.9s ease-in-out,
        border-bottom 0.9s ease-in-out,
        border 0.9s ease-in-out,
        border 0.9s ease-in-out,
        background-color 0.9s ease-in-out,
        background-color 0.9s ease-in-out;
    }

    input:focus {
      outline: none;
    }

    ::placeholder {
      color: variables.$placeholder-color;
    }

    .search-form-button {
      background-color: transparent;
      border: none;
      margin: 0 10px;
    }
  }

  .dropdown {
    position: absolute;
    overflow: auto;
    left: 50%;
    margin-top: variables.$xxs-margin;
    margin-left: -50%;
    max-height: 298px;
    width: 100%;
    border-radius: variables.$border-radius;
    background-color: variables.$primary-medium-color;

    .surname {
      padding-left: variables.$xxxs-margin;
      font-weight: 500;
    }

    .dates {
      padding-left: variables.$xxxs-margin;
      font-size: 1rem;
    }

    p {
      margin: 0;
      padding: variables.$xxs-margin variables.$xs-margin;
      transition:
        color 0.9s ease-in-out,
        background-color 0.9s ease-in-out;

      &:hover {
        color: variables.$secondary-color;
        background-color: variables.$primary-light-color;
      }
    }

    @include variables.tablet {
      max-height: 350px;
    }

    @include variables.desktop {
      max-height: 335px;
    }
  }
}

.hidden {
  opacity: 0;
}
