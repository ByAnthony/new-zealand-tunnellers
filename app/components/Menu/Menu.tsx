'use client'

import { useEffect, useState } from 'react';

import STYLES from './Menu.module.scss';
import { Tunneller } from '../../../app/types/tunnellers';

type Props = {
  tunnellers: Tunneller[];
};

export function Menu({tunnellers}: Props) {
  const [filteredTunnellers, setFilteredTunnellers] = useState<Tunneller[]>([]);
  const [prevScrollPosition, setPrevScrollPosition] = useState(0);
  const [menuVisible, setMenuVisible] = useState(true);

  useEffect(() => {
    const handleScroll = () => {
      const currentScrollPosition = window.scrollY;

      if (prevScrollPosition > currentScrollPosition) {
        setMenuVisible(true);
      } else {
        setMenuVisible(false);
      }

      setPrevScrollPosition(currentScrollPosition);
    };

    window.addEventListener('scroll', handleScroll);
    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, [prevScrollPosition]);

  const handleSearch = (search: string) => {
    const searchParts = search.toLowerCase().split(' ');

    setFilteredTunnellers(tunnellers.filter((tunneller: Tunneller) => {
      const fullName = tunneller.fullName?.toLowerCase() || '';
      return searchParts.every(part => fullName.includes(part));
    }));  
  };

  const capitalizeName = (name: string) => {
    return name.replace(/\b\w/g, char => char.toUpperCase());
  };

  return (
    <>
      <div data-testid="menu" className={`${STYLES.menu} ${menuVisible ? '' : STYLES.hidden}`}>
        <a href="/" className={STYLES.logo} aria-label="Go to the Homepage"><img src="/nzt_logo.png" alt="" /></a>
        <div className={STYLES['search-form']}>
          <form>
            <input type="text" id="search" name="search" placeholder="Search for a Tunneller" onChange={(event) => handleSearch(event.target.value)} />
          </form>
            <img src="/searching_hover.png" className={STYLES['search-form-button']} alt="" height={18} />
        </div>
      </div>
      <div className={STYLES['drop-down']}>
        <ul>
          {filteredTunnellers.map((tunneller, index) => (
            <li key={index}><a href={`/tunnellers/${tunneller.id}`}>{capitalizeName(tunneller.fullName)}</a></li>
          ))}
        </ul>
      </div>
    </>
  );
}
