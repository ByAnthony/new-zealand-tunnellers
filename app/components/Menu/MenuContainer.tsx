import { getTunnellers } from "../../../utils/database/getEndpoint";
import { Menu } from "./Menu";

export async function MenuContainer() {
  const tunnellers = await getTunnellers();

  return <Menu tunnellers={tunnellers} />;
}
