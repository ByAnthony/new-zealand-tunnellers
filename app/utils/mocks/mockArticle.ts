import { AboutUsArticle } from "../../types/article";

export const mockArticle: AboutUsArticle = {
  id: "about-us",
  title: "Our Story",
  section: [
    { title: "Title 1", text: "This is the first paragraph of our story." },
    { title: "Title 2", text: "This is the second paragraph of our story." },
  ],
  image: [
    {
      file: "our-story.jpg",
      title: null,
      photographer: null,
      reference: null,
      alt: "A picture of our team",
    },
  ],
};
