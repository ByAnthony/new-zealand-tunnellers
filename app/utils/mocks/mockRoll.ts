import { Tunneller } from '@/app/types/roll';
import { mockId } from './mockProfile';

const johnDoe = {
  forename: 'John',
  surname: 'Doe',
};

const alexanderDriscott = {
  forename: 'Alexander',
  surname: 'Driscott',
};

const williamRight = {
  forename: 'William',
  surname: 'Right',
};

const robertLang = {
  forename: 'Robert',
  surname: 'Lang',
};

export const mockDetails: Tunneller = {
  id: mockId,
  name: johnDoe,
  birthDate: '1886',
  deathDate: '1952',
};

export const mockDetailsList: Tunneller[] = [
  mockDetails,
  {
    ...mockDetails,
    id: 1,
    name: alexanderDriscott,
  },
];

export const mockRoll: Record<string, Tunneller[]> = {
  D: [
    mockDetails,
    {
      ...mockDetails,
      id: 1,
      name: alexanderDriscott,
    },
  ],
  L: [
    {
      ...mockDetails,
      id: 2,
      name: robertLang,
    },
  ],
  R: [
    {
      ...mockDetails,
      id: 3,
      name: williamRight,
    },
  ],
};
