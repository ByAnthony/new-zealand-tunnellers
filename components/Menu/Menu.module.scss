@use "app/variables.module.scss";

.menu {
  position: -webkit-sticky;
  position: sticky;
  top: variables.$m-margin;
  display: flex;
  justify-content: center;
  align-items: center;
  transition: opacity 0.9s ease-in-out;

  @include variables.component-margin;

  @include variables.tablet {
    top: variables.$m-margin;
  }

  @include variables.desktop {
    margin: 0 auto;
    top: variables.$base-margin;
    max-width: variables.$inner-width;
  }

  .logo {
    display: flex;
    align-items: center;
    height: variables.$base-margin;
    margin-right: variables.$xs-margin;
  }

  .search-form-container {
    position: relative;
    width: 100%;
  }

  .search-form {
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: variables.$base-margin;
    border: 1px solid variables.$primary-light-color;
    border-radius: variables.$border-radius;
    background-color: variables.$primary-medium-color;

    input {
      width: 100%;
      box-sizing: border-box;
      width: 100%;
      height: variables.$base-margin;
      padding: 0 variables.$xs-margin;
      font-family: variables.$main-font;
      font-weight: 300;
      font-size: 1rem;
      color: variables.$text-color;
      border: none;
      border-top-left-radius: variables.$border-radius;
      border-bottom-left-radius: variables.$border-radius;
      background-color: variables.$primary-medium-color;
      transition:
        border-bottom 0.9s ease-in-out,
        border-bottom 0.9s ease-in-out,
        border 0.9s ease-in-out,
        border 0.9s ease-in-out,
        background-color 0.9s ease-in-out,
        background-color 0.9s ease-in-out;
    }

    input:focus {
      outline: none;
    }

    ::placeholder {
      color: variables.$placeholder-color;
    }

    .search-form-button {
      background-color: transparent;
      border: none;
      margin: 0 10px;
    }
  }

  .dropdown {
    position: absolute;
    overflow: auto;
    display: flex;
    flex-direction: column;
    left: 50%;
    margin-top: variables.$xxs-margin;
    margin-left: -50%;
    width: 100%;
    border-radius: variables.$border-radius;
    border: 1px solid variables.$primary-light-color;
    background-color: variables.$primary-medium-color;

    .surname {
      font-weight: 500;
    }

    .dates {
      font-size: 1rem;
    }

    ul {
      overflow: scroll;
      max-height: 298px;

      @include variables.tablet {
        max-height: 350px;
      }

      @include variables.desktop {
        max-height: 328px;
      }
    }

    p {
      margin: 0;
      padding: variables.$xxs-margin variables.$xs-margin;
      transition:
        color 0.9s ease-in-out,
        background-color 0.9s ease-in-out;

      &:hover {
        color: variables.$secondary-color;
        background-color: variables.$primary-light-color;
      }
    }
  }
}

.tunnellers-link {
  margin-top: -17px;
  padding: variables.$xs-margin variables.$xs-margin variables.$xs-margin
    variables.$xs-margin;
  color: variables.$secondary-color;
  font-weight: 500;
  border-top: 1px solid variables.$primary-light-color;
  background-color: variables.$primary-dark-color;
  transition:
    color 0.9s ease-in-out,
    background-color 0.9s ease-in-out;

  &:hover {
    color: variables.$secondary-color;
    background-color: variables.$primary-light-color;
  }
}

.tunnellers-link-display {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}

.arrow {
  line-height: 20px;
}

.hidden {
  opacity: 0;
  transition: opacity 0.9s ease-in-out;
}
