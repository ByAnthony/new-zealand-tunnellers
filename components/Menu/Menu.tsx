"use client";

import { useEffect, useRef, useState } from "react";
import Image from "next/image";
import Link from "next/link";
import { usePathname } from "next/navigation";

import { TunnellerWithFullNameData } from "@/types/tunnellers";
import { displayBiographyDates } from "@/utils/helpers/roll";

import STYLES from "./Menu.module.scss";

type Props = {
  tunnellers: TunnellerWithFullNameData[];
};

export function Menu({ tunnellers }: Props) {
  const divRef = useRef<HTMLUListElement>(null);
  const pathname = usePathname();

  const [filteredTunnellers, setFilteredTunnellers] = useState<
    TunnellerWithFullNameData[]
  >([]);
  const [dropdownVisible, setDropdownVisible] = useState(false);

  const handleClickOutside = (event: MouseEvent) => {
    if (divRef.current && !divRef.current.contains(event.target as Node)) {
      setDropdownVisible(false);
    }
  };

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);

  const handleSearch = (search: string) => {
    const searchParts = search.toLowerCase().split(" ");

    setFilteredTunnellers(
      search.length > 0
        ? tunnellers.filter((tunneller: TunnellerWithFullNameData) => {
            const fullName = tunneller.fullName?.toLowerCase();
            return searchParts.every((part) => fullName.includes(part));
          })
        : [],
    );
    setDropdownVisible(search.length > 0 ? true : false);
  };

  const showDropdown = () => {
    setDropdownVisible(true);
  };

  return (
    <div key={pathname} data-testid="menu" className={`${STYLES.menu} fade-in`}>
      <Link href="/" className={STYLES.logo} aria-label="Go to the Homepage">
        <Image
          src="/nzt_logo.png"
          alt="New Zealand Tunnellers Wesbite - Home"
          width={30}
          height={30}
          priority={true}
        />
      </Link>
      <div className={STYLES["search-form-container"]}>
        <div className={STYLES["search-form"]} onClick={showDropdown}>
          <input
            type="text"
            id="search"
            placeholder="Search for a Tunneller"
            onChange={(event) => handleSearch(event.target.value)}
          />
          <Image
            src="/search.png"
            alt="Type a name to search for a Tunneller"
            width={20}
            height={20}
            className={STYLES["search-form-button"]}
          />
        </div>
        {dropdownVisible && (
          <ul className={STYLES.dropdown} ref={divRef}>
            {filteredTunnellers.map((tunneller, index) => (
              <li key={index}>
                <Link
                  href={`/tunnellers/${tunneller.id}`}
                  aria-label={`See ${tunneller.forename} ${tunneller.surname} profile`}
                >
                  <p>
                    {tunneller.forename}
                    <span className={STYLES.surname}>
                      {` ${tunneller.surname} `}
                    </span>
                    <span className={STYLES.dates}>
                      (
                      {displayBiographyDates(
                        tunneller.birthYear,
                        tunneller.deathYear,
                      )}
                      )
                    </span>
                  </p>
                </Link>
              </li>
            ))}
          </ul>
        )}
      </div>
    </div>
  );
}
