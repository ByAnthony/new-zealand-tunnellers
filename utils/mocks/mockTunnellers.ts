import { Tunneller, TunnellerWithFullNameData } from "@/types/tunnellers";

const mockTunnellerData1: TunnellerWithFullNameData = {
  id: 1,
  forename: "John",
  surname: "Doe",
  birthYear: "1886",
  deathYear: "1952",
  fullName: "John Doe",
};

const mockTunnellerData2: TunnellerWithFullNameData = {
  id: 2,
  forename: "Biff",
  surname: "Tanen",
  birthYear: "1897",
  deathYear: null,
  fullName: "William Smith",
};

const mockTunnellerData3: TunnellerWithFullNameData = {
  id: 3,
  forename: "Emmett",
  surname: "Brown",
  birthYear: null,
  deathYear: "1935",
  fullName: "Emmett Brown",
};

const mockTunnellerData4: TunnellerWithFullNameData = {
  id: 4,
  forename: "Marty",
  surname: "McFly",
  birthYear: null,
  deathYear: null,
  fullName: "Marty McFly",
};

export const mockTunnellersData: TunnellerWithFullNameData[] = [
  mockTunnellerData1,
  mockTunnellerData2,
  mockTunnellerData3,
  mockTunnellerData4,
];

const mockTunneller1: Tunneller = {
  id: 1,
  name: { forename: "John", surname: "Doe" },
  birthYear: "1886",
  deathYear: "1952",
  fullName: "John Doe",
};

const mockTunneller2: Tunneller = {
  id: 2,
  name: { forename: "Biff", surname: "Tanen" },
  birthYear: "1897",
  deathYear: null,
  fullName: "William Smith",
};

const mockTunneller3: Tunneller = {
  id: 3,
  name: { forename: "Emmett", surname: "Brown" },
  birthYear: null,
  deathYear: "1935",
  fullName: "Emmett Brown",
};

const mockTunneller4: Tunneller = {
  id: 4,
  name: { forename: "Marty", surname: "McFly" },
  birthYear: null,
  deathYear: null,
  fullName: "Marty McFly",
};

export const mockTunnellers: Record<string, Tunneller[]> = {
  B: [mockTunneller3],
  D: [mockTunneller1],
  M: [mockTunneller4],
  S: [mockTunneller2],
};
