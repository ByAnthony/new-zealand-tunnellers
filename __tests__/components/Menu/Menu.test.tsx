import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import { Menu } from "../../../app/components/Menu/Menu";
import { mockTunnellersData } from "../../mocks/mockTunnellers";
import { doc } from "prettier";

describe("Menu", () => {
  test("matches the snapshot", () => {
    const { asFragment } = render(<Menu tunnellers={mockTunnellersData} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test("renders the component correctly", () => {
    render(<Menu tunnellers={mockTunnellersData} />);

    const nextButton = screen.getByRole("link", {
      name: "Go to the Homepage",
    });
    expect(nextButton).toHaveAttribute("href", "/");

    const search = screen.getByRole("textbox");
    expect(search).toBeInTheDocument();
    expect(search).toHaveAttribute("placeholder", "Search for a Tunneller");
  });

  test("can input a name", () => {
    render(<Menu tunnellers={mockTunnellersData} />);

    const search = screen.getByRole("textbox");
    fireEvent.click(search);
    fireEvent.change(search, {
      target: { value: "John Doe" },
    });

    expect(screen.getByRole("list")).toBeInTheDocument();
    expect(screen.getByText("John")).toBeInTheDocument();
    expect(screen.getByText("Doe")).toBeInTheDocument();
    expect(screen.getByText("Doe")).toHaveClass("surname");
    expect(screen.getByText("(1886-1952)")).toBeInTheDocument();
    expect(screen.getAllByRole("link")[1]).toHaveAttribute(
      "href",
      "/tunnellers/1",
    );
  });

  test("can close the dropdown with outside click", () => {
    render(<Menu tunnellers={mockTunnellersData} />);

    const search = screen.getByRole("textbox");
    fireEvent.click(search);
    fireEvent.change(search, {
      target: { value: "John Doe" },
    });

    expect(screen.getByRole("list")).toBeInTheDocument();

    fireEvent.mouseDown(document.body);
    expect(screen.queryByRole("list")).not.toBeInTheDocument();
  });

  test("can clear name and close the dropdown", () => {
    render(<Menu tunnellers={mockTunnellersData} />);

    const search = screen.getByRole("textbox");
    fireEvent.click(search);
    fireEvent.change(search, {
      target: { value: "John Doe" },
    });

    expect(screen.getByRole("list")).toBeInTheDocument();

    fireEvent.click(search);
    fireEvent.change(search, {
      target: { value: "" },
    });
    expect(screen.queryByRole("list")).not.toBeInTheDocument();
  });
});
