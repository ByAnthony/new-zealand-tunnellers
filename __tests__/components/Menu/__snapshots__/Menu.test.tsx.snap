// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="menu"
    data-testid="menu"
  >
    <a
      aria-label="Go to the Homepage"
      class="logo"
      href="/"
    >
      <img
        alt="New Zealand Tunnellers Wesbite - Home"
        data-nimg="1"
        decoding="async"
        height="30"
        loading="lazy"
        src="/_next/image?url=%2Fnzt_logo.png&w=64&q=75"
        srcset="/_next/image?url=%2Fnzt_logo.png&w=32&q=75 1x, /_next/image?url=%2Fnzt_logo.png&w=64&q=75 2x"
        style="color: transparent;"
        width="30"
      />
    </a>
    <div
      class="search-form-container"
    >
      <div
        class="search-form"
      >
        <input
          id="search"
          placeholder="Search for a Tunneller"
          type="text"
        />
        <img
          alt="Type a name to search for a Tunneller"
          class="search-form-button"
          data-nimg="1"
          decoding="async"
          height="20"
          loading="lazy"
          src="/_next/image?url=%2Fsearch.png&w=48&q=75"
          srcset="/_next/image?url=%2Fsearch.png&w=32&q=75 1x, /_next/image?url=%2Fsearch.png&w=48&q=75 2x"
          style="color: transparent;"
          width="20"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
